rs/registry/canister/src/mutations/node_management/do_replace_operator.rs

```diff
@@ -0,0 +1,628 @@
+use ic_registry_canister_api::UpdateNodeOperatorPayload;
+use ic_registry_keys::{make_node_operator_record_key, make_node_record_key};
+use ic_registry_transport::update;
+use ic_types::{NodeId, PrincipalId};
+use prost::Message;
+
+impl Registry {
+    /// Replaces the node's operator id with a new operator id
+    /// that is in the same data center and is related to the
+    /// same node provider.
+    ///
+    /// Expected caller of this function has to be a principal
+    /// related to a node provider. All other principals will
+    /// be rejected.
+    ///
+    /// It is expected that all the nodes currently have set
+    /// `old_node_operator` as their node operator.
+    ///
+    /// Both `old_node_operator` and `new_node_operator` must
+    /// belong to the same node provider, who is the caller,
+    /// and must be within the same data center.
+    pub fn do_replace_operator(
+        &mut self,
+        payload: UpdateNodeOperatorPayload,
+    ) -> Result<(), String> {
+        let caller_id = dfn_core::api::caller();
+        println!(
+            "{}do_replace_operator: {:?} caller: {:?}",
+            LOG_PREFIX, payload, caller_id
+        );
+
+        self.do_replace_operator_with_caller(payload, caller_id)
+    }
+
+    fn do_replace_operator_with_caller(
+        &mut self,
+        payload: UpdateNodeOperatorPayload,
+        caller_id: PrincipalId,
+    ) -> Result<(), String> {
+        payload
+            .validate()
+            .map_err(|e| format!("{}do_replace_operator: {}", LOG_PREFIX, e))?;
+
+        let node_ids = payload.node_ids.clone().unwrap();
+        let old_operator_id = payload.old_operator_id.unwrap();
+        let new_operator_id = payload.new_operator_id.unwrap();
+
+        // Fetch operator records and extract
+        // the records related to the provided
+        // payload.
+        let operators = self.maybe_fetch_operators_for_provider(&caller_id)?;
+
+        let (old_operator_record, new_operator_record) = self.maybe_find_operator_records(
+            operators,
+            &caller_id,
+            &old_operator_id,
+            &new_operator_id,
+        )?;
+
+        let mut valid_mutations: Vec<_> = self
+            .maybe_fetch_nodes_to_update(&node_ids, &new_operator_id, &old_operator_id)?
+            .into_iter()
+            .map(|(node_id, node_record)| {
+                let node_key = make_node_record_key(node_id);
+                let updated_node_record = NodeRecord {
+                    node_operator_id: new_operator_id.as_slice().to_vec(),
+                    ..node_record
+                };
+                update(node_key, updated_node_record.encode_to_vec())
+            })
+            .collect();
+
+        // Nothing should be done.
+        if valid_mutations.is_empty() {
+            return Ok(());
+        }
+
+        let node_mutations = valid_mutations.len() as u64;
```

Comment:
No action required.

I think it is better to reserve plural for naming collections, not cardinalities. However, I think ppl use it for both; hence, NAR. What I would suggest in this particular situation is just inline foo.len() as u64. Yes, it's a little bit more verbose, but the difference is pretty small. IMHO, the additional indirection is not worth it in this case.

Reply:
I can't inline it just because I later modify the same list meaning that `valid_mutations.len()` changes. I could rely on expected math and then subtract, but I found this to be a little bit cleaner.